ambry-network/src/main/java/com.github.ambry.network/ConnectionTracker.java

```diff
@@ -56,21 +61,85 @@ boolean mayCreateNewConnection(String host, Port port) {
+   * @param host the hostname.
+   * @param port the port number.
+   * @param warmUpPercentage percentage of max connections to this (host, port) that should be kept ready for use.
+   */
+  void enableConnectionWarmUp(String host, Port port, int warmUpPercentage) {
+    HostPortPoolManager hostPortPoolManager = getHostPortPoolManager(host, port);
+    hostPortPoolManager.setLowWatermark(warmUpPercentage * hostPortPoolManager.poolLimit / 100);
+    if (!hostPortPoolManager.hasReachedLowWatermark()) {
+      poolManagersBelowWatermark.add(hostPortPoolManager);
+    }
+  }
+
+  /**
+   * For (host, port) pools that are below the connection watermark, initiate new connections to each host until they
+   * meet
+   * @param connectionFactory the {@link ConnectionFactory} for interfacing with the networking layer.
+   * @return the number of connections initiated.
+   */
+  int replenishConnections(ConnectionFactory connectionFactory) {
+    int connectionsInitiated = 0;
+    Iterator<HostPortPoolManager> iter = poolManagersBelowWatermark.iterator();
+    while (iter.hasNext()) {
+      HostPortPoolManager poolManager = iter.next();
+      try {
+        while (!poolManager.hasReachedLowWatermark()) {
+          connectAndTrack(connectionFactory, poolManager.host, poolManager.port, false);
+          connectionsInitiated++;
+        }
+        iter.remove();
+      } catch (IOException e) {
+        LOGGER.warn("Encountered exception while replenishing connections to {}:{}.", poolManager.host,
+            poolManager.port.getPort(), e);
+      }
+    }
+    return connectionsInitiated;
+  }
+
+  /**
+   * Initiate a new connection using the provided {@link ConnectionFactory} and start tracking a new connection id
+   * associated with the given host and port. Note that this connection will not be made available for checking out
+   * until a {@link #checkInConnection} is called on it.
+   * @param connectionFactory the {@link ConnectionFactory} for interfacing with the networking layer.
+   * @param host the host to which this connection belongs.
+   * @param port the port on the host to which this connection belongs.
+   * @return the connection id of the connection returned by {@link ConnectionFactory#connect}.
+   */
+  String connectAndTrack(ConnectionFactory connectionFactory, String host, Port port) throws IOException {
+    return connectAndTrack(connectionFactory, host, port, true);
+  }
+
+  /**
+   * @see #connectAndTrack(ConnectionFactory, String, Port)
+   * @param connectionFactory the {@link ConnectionFactory} for interfacing with the networking layer.
    * @param host the host to which this connection belongs.
    * @param port the port on the host to which this connection belongs.
-   * @param connId the connection id of the connection.
+   * @param editPoolManagersBelowWatermark true to allow this method to edit {@link #poolManagersBelowWatermark}.
```

Comment:
Do we ever add to `poolManagersBelowWatermark` when this is true or is it always removing?  If the latter perhaps we should call this `shouldRemovePoolManagersBelowWatermark`, if former `shouldEditPoolManagersBelowWatermark` (this is actually kind of a weird parameter in general since it's exposing the management of an internal detail (the `poolManagersBelowWatermark` map) to the user.  Can you explain its necessity in the context of its use?)

Reply:
You are right, this method only removes from that set. 

This is a private helper method meant to allow `connectAndTrack` and `replenishConnections` to share code. `Set::remove()` cannot be called in `replenishConnections` because it uses an `Iterator` so I added that flag to allow the two methods to both work correctly.

As an alternative, I can just duplicate the small bit of shared code and get rid of this method to reduce the confusion.